File: crates/types/src/performance_monitoring.rs
use rust_decimal::Decimal;
use serde::{Deserialize, Serialize};

pub const MAX_SERIES_POINTS: usize = 720;

#[derive(Clone, Copy, Debug, PartialEq, Eq, Deserialize, Serialize)]
#[serde(rename_all = "lowercase")]
pub enum PerformanceMonitoringRange {
    Today,
    #[serde(rename = "7d")]
    SevenDays,
    #[serde(rename = "30d")]
    ThirtyDays,
    All,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize)]
#[serde(rename_all = "lowercase")]
pub enum SnapshotGranularity {
    Minute,
    Hour,
    Day,
}

#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
pub struct PerformanceMonitoringOverviewRequest {
    #[serde(default)]
    pub range: PerformanceMonitoringRange,
}

#[derive(Clone, Debug, PartialEq, Serialize)]
pub struct PerformanceMonitoringOverviewResponse {
    pub range: PerformanceMonitoringRange,
    pub effective_range: EffectiveTimeRange,
    pub bucket_granularity: SnapshotGranularity,
    pub max_series_points: usize,
    pub status: SnapshotDataStatus,
    pub series: Vec<PerformanceSnapshotPoint>,
}

#[derive(Clone, Debug, PartialEq, Serialize)]
pub struct PerformanceMonitoringRealtimeResponse {
    pub snapshot: Option<PerformanceSnapshotPoint>,
    pub host: HostRealtimeMetrics,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize)]
pub struct EffectiveTimeRange {
    pub started_at: String,
    pub ended_at: String,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum SnapshotDataStatus {
    Ready,
    EmptySnapshot,
}

#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
pub struct PerformanceSnapshotMetrics {
    pub core: CoreRequestMetrics,
    pub llm: LlmBusinessMetrics,
    pub network: NetworkConnectionMetrics,
    pub host: HostResourceMetrics,
}

#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
pub struct CoreRequestMetrics {
    pub request_count: i64,
    pub qps: f64,
    pub concurrent_requests: i64,
    pub success_rate: f64,
    pub error_rate: f64,
    pub timeout_rate: f64,
    pub rate_limited_count: i64,
    pub server_error_count: i64,
    pub p50_latency_ms: Option<i64>,
    pub p95_latency_ms: Option<i64>,
    pub p99_latency_ms: Option<i64>,
    pub p50_ttft_ms: Option<i64>,
    pub p95_ttft_ms: Option<i64>,
    pub p99_ttft_ms: Option<i64>,
    pub retry_count: i64,
    pub circuit_breaker_count: i64,
    pub stream_request_count: i64,
}

#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
pub struct LlmBusinessMetrics {
    pub prompt_tokens: i64,
    pub completion_tokens: i64,
    pub total_tokens: i64,
    pub tokens_per_request: f64,
    pub tokens_per_second: f64,
    pub model_distribution: Vec<MetricDimension>,
    pub provider_distribution: Vec<MetricDimension>,
    pub failover_count: i64,
    pub cache_hit_rate: f64,
    #[serde(with = "rust_decimal::serde::float")]
    pub cost: Decimal,
    pub quota_limited_count: i64,
}

#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
pub struct NetworkConnectionMetrics {
    pub inbound_bytes: i64,
    pub outbound_bytes: i64,
    pub inbound_bandwidth_bytes_per_second: f64,
    pub outbound_bandwidth_bytes_per_second: f64,
    pub current_connections: Option<i64>,
    pub new_connections_per_second: Option<f64>,
    pub tcp_total: Option<i64>,
    pub tcp_time_wait: Option<i64>,
    pub tcp_established: Option<i64>,
    pub tcp_close_wait: Option<i64>,
    pub retransmits: Option<i64>,
    pub packet_loss: Option<i64>,
    pub status: MetricSupportStatus,
}

#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
pub struct HostResourceMetrics {
    pub cpu_usage_percent: Option<f64>,
    pub load_average_1m: Option<f64>,
    pub load_average_5m: Option<f64>,
    pub load_average_15m: Option<f64>,
    pub memory_rss_bytes: Option<i64>,
    pub memory_usage_bytes: Option<i64>,
    pub disk_total_bytes: Option<i64>,
    pub disk_available_bytes: Option<i64>,
    pub disk_read_bytes_per_second: Option<f64>,
    pub disk_write_bytes_per_second: Option<f64>,
    pub file_descriptors: Option<i64>,
    pub threads: Option<i64>,
    pub processes: Option<i64>,
    pub status: MetricSupportStatus,
}

#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
pub struct MetricDimension {
    pub name: String,
    pub count: i64,
}

#[derive(Clone, Copy, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum MetricSupportStatus {
    #[default]
    Unsupported,
    Ready,
}

#[derive(Clone, Debug, PartialEq, Serialize)]
pub struct PerformanceSnapshotPoint {
    pub bucket_started_at: String,
    pub bucket_ended_at: String,
    pub metrics: PerformanceSnapshotMetrics,
}

#[derive(Clone, Debug, PartialEq, Serialize)]
pub struct HostRealtimeMetrics {
    pub collected_at: String,
    pub metrics: HostResourceMetrics,
}

impl Default for PerformanceMonitoringRange {
    fn default() -> Self {
        Self::Today
    }
}

impl SnapshotGranularity {
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::Minute => "minute",
            Self::Hour => "hour",
            Self::Day => "day",
        }
    }

    pub const fn bucket_seconds(self) -> i64 {
        match self {
            Self::Minute => 60,
            Self::Hour => 3_600,
            Self::Day => 86_400,
        }
    }
}

impl TryFrom<&str> for SnapshotGranularity {
    type Error = String;

    fn try_from(value: &str) -> Result<Self, Self::Error> {
        match value {
            "minute" => Ok(Self::Minute),
            "hour" => Ok(Self::Hour),
            "day" => Ok(Self::Day),
            _ => Err(format!("unsupported snapshot granularity: {value}")),
        }
    }
}

File: apps/hook_frontend/src/types/performance-monitoring.ts
export type PerformanceMonitoringRange = 'realtime' | 'today' | '7d' | '30d' | 'all';
export type SnapshotGranularity = 'minute' | 'hour' | 'day';
export type SnapshotDataStatus = 'ready' | 'empty_snapshot';
export type MetricSupportStatus = 'ready' | 'unsupported';

export type EffectiveTimeRange = {
  started_at: string;
  ended_at: string;
};

export type MetricDimension = {
  name: string;
  count: number;
};

export type CoreRequestMetrics = {
  request_count: number;
  qps: number;
  concurrent_requests: number;
  success_rate: number;
  error_rate: number;
  timeout_rate: number;
  rate_limited_count: number;
  server_error_count: number;
  p50_latency_ms?: number | null;
  p95_latency_ms?: number | null;
  p99_latency_ms?: number | null;
  p50_ttft_ms?: number | null;
  p95_ttft_ms?: number | null;
  p99_ttft_ms?: number | null;
  retry_count: number;
  circuit_breaker_count: number;
  stream_request_count: number;
};

export type LlmBusinessMetrics = {
  prompt_tokens: number;
  completion_tokens: number;
  total_tokens: number;
  tokens_per_request: number;
  tokens_per_second: number;
  model_distribution: MetricDimension[];
  provider_distribution: MetricDimension[];
  failover_count: number;
  cache_hit_rate: number;
  cost: number;
  quota_limited_count: number;
};

export type NetworkConnectionMetrics = {
  inbound_bytes: number;
  outbound_bytes: number;
  inbound_bandwidth_bytes_per_second: number;
  outbound_bandwidth_bytes_per_second: number;
  current_connections?: number | null;
  new_connections_per_second?: number | null;
  tcp_total?: number | null;
  tcp_time_wait?: number | null;
  tcp_established?: number | null;
  tcp_close_wait?: number | null;
  retransmits?: number | null;
  packet_loss?: number | null;
  status: MetricSupportStatus;
};

export type HostResourceMetrics = {
  cpu_usage_percent?: number | null;
  load_average_1m?: number | null;
  load_average_5m?: number | null;
  load_average_15m?: number | null;
  memory_rss_bytes?: number | null;
  memory_usage_bytes?: number | null;
  disk_total_bytes?: number | null;
  disk_available_bytes?: number | null;
  disk_read_bytes_per_second?: number | null;
  disk_write_bytes_per_second?: number | null;
  file_descriptors?: number | null;
  threads?: number | null;
  processes?: number | null;
  status: MetricSupportStatus;
};

export type PerformanceSnapshotMetrics = {
  core: CoreRequestMetrics;
  llm: LlmBusinessMetrics;
  network: NetworkConnectionMetrics;
  host: HostResourceMetrics;
};

export type PerformanceSnapshotPoint = {
  bucket_started_at: string;
  bucket_ended_at: string;
  metrics: PerformanceSnapshotMetrics;
};

export type PerformanceMonitoringOverviewResponse = {
  range: PerformanceMonitoringRange;
  effective_range: EffectiveTimeRange;
  bucket_granularity: SnapshotGranularity;
  max_series_points: number;
  status: SnapshotDataStatus;
  series: PerformanceSnapshotPoint[];
};

export type HostRealtimeMetrics = {
  collected_at: string;
  metrics: HostResourceMetrics;
};

export type PerformanceMonitoringRealtimeResponse = {
  snapshot?: PerformanceSnapshotPoint | null;
  host: HostRealtimeMetrics;
};
